// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

syntax = "proto3";

package google.crypto.tink;

import "proto/common.proto";

option go_package = "github.com/google/tink/proto/tpm_common_go_proto";

message TpmObjectAuthPolicy {
  bool password_required = 1;

  // Platform Configuration Registers (PCRs) are a set of fixed-sized registers
  // in the TPM. Arbitrary values can't be directly written into this PCRs, only
  // extended:
  // ExtendPCR(pcr_idx, content) | PCR[pcr_idx] = HASH_ALG( PCR[i] | content )
  // This provides certain defence against forgery. With this property, PCRs are
  // used to catpure the state of a system (go/tpm-rtms).
  message Pcr {
    uint32 num = 1;
    HashType hash = 2;
  }
  repeated Pcr pcrs = 2;
}

// Defines the public parameters needed to re-create a Storage Primary Key in
// the TPM. The public_template is a marshaled TPM2B_PUBLIC structure
// (https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part2_Structures_pub.pdf#page=160).
message TpmPrimaryKey {
  bytes public_template = 1;
}

// A Hierarchy is a collection of entities that are related and managed as a
// group. All primary objects in the TPM are places under one of the following
// hierarchies.
enum TpmHierarchy {
  HIERARCHY_UNSPECIFIED = 0;
  // TPM (reserved handle) hierarchy for platform owner.
  HIERARCHY_OWNER = 1;
  // TPM (reserved handle) hierarchy for platform manufacter.
  HIERARCHY_PLATFORM = 2;
  // TPM (reserved handle) hierarchy for privacy administration.
  HIERARCHY_ENDORSEMENT = 3;
}
